### ISSUE TITLE
java.nio.channels.NotYetConnectedException raised even after IO.select

### ISSUE THREAD

**Issue by Zapotek** [jruby/jruby#1694]
There seem to be a bug with JRuby (`jruby 1.7.12 (2.0.0p195) 2014-04-15 643e292 on Java HotSpot(TM) 64-Bit Server VM 1.7.0_55-b13 [linux-amd64]`) due to which a non-blocking connect operation successfully takes place but that fact doesn't register internally -- or looks like it doesn't.

MRI 2.1.1 completes without issue while JRuby fails with `java.nio.channels.NotYetConnectedException` even after `IO.select` reports the socket ready for writing (or reading, same issue exists).

``` ruby
require 'socket'

def connect( host, port )
    socket = Socket.new(
        Socket::Constants::AF_INET,
        Socket::Constants::SOCK_STREAM,
        Socket::Constants::IPPROTO_IP
    )

    begin
        socket.connect_nonblock( Socket.sockaddr_in( port, host ) )

    # Will throw IO::EINPROGRESSWaitWritable, as expected.
    rescue IO::WaitReadable, IO::WaitWritable, Errno::EINPROGRESS
    end

    socket
end

address = ['127.0.0.1', 7331]

server = TCPServer.new( *address )

# The connection will indeed be established as 'accepted' will be printed to
# the console.
Thread.new do
    server.accept
    p 'accepted'
end

client = connect( *address )

p IO.select( nil, [client] )
# => [[], [#<Socket:fd 17>], []]

client.write_nonblock 'test'
# Will fail with java.nio.channels.NotYetConnectedException
# #write, #read, and #read_nonblock will fail with the same error as well.
```

**Comment by headius** [comment]
This is still an issue on 9.1.

**Comment by deep-thought-42** [comment]
Hi Charles, we are also facing the same issue. Here is the complete stack trace for reference.
https://gist.github.com/deep-thought-42/3888031f49f92ed65f8c8f557276bd9b

JRuby 9.1.0.0
java version "1.8.0_65"
Java(TM) SE Runtime Environment (build 1.8.0_65-b17)
Java HotSpot(TM) 64-Bit Server VM (build 25.65-b01, mixed mode)

**Comment by enebo** [comment]
Added grist for the mill.  Virtually all of the test_nonblock.rb tests in MRI test suite throw this same error message from nio.  I am guessing something is not quite setup and in some cases I think it is just expecting a Ruby error and we are not connecting them up.  I know @headius had a fix for that last bit.

**Comment by headius** [comment]
A recent fix on master (as part of #6311) improved our handling of the `NotYetConnectedException` by re-raising it as ENOTCONN. That doesn't resolve this issue (if the original script is expected to work) but it does improve the error:

```
$ jruby blah.rb
"accepted"
[[], [#<Socket:fd 22>], []]
Errno::ENOTCONN: Socket is not connected - No message available
  write_nonblock at org/jruby/RubyIO.java:1366
          <main> at blah.rb:36
```

**Comment by headius** [comment]
On second thought I'm not sure there's much we can fix here, and maybe it's not a big deal.

With the `select` loop in place, this does work properly; we detect that the connection is "pending" and ask the JDK classes to complete the connection logic.

MRI is able to go straight into using the connection only because it does not maintain any other state surrounding the socket descriptor.

We may behave like CRuby if (when?) we manage to support the pure-Ruby FFI-based socket library from Rubinius, but until then this is simply a requirement of the Java socket API. Do the select loop, call `connect_nonblock` until it succeeds, and it will always work.
